File: go/graph.go
package sssp

// Graph is a directed graph with non-negative edge weights.
// Vertices are 0..VertexCount-1.
type Graph struct {
	outEdges  [][]edge
	edgeCount int
}

type edge struct {
	to     int
	weight float64
}

// NewGraph creates a graph with n vertices.
func NewGraph(n int) *Graph {
	return &Graph{
		outEdges: make([][]edge, n),
	}
}

// VertexCount returns the number of vertices.
func (g *Graph) VertexCount() int {
	return len(g.outEdges)
}

// EdgeCount returns the number of edges.
func (g *Graph) EdgeCount() int {
	return g.edgeCount
}

// AddEdge adds a directed edge (from, to) with the given weight.
func (g *Graph) AddEdge(from, to int, weight float64) {
	if weight < 0 {
		panic("edge weights must be non-negative")
	}
	g.outEdges[from] = append(g.outEdges[from], edge{to, weight})
	g.edgeCount++
}

// OutEdges returns the outgoing edges from vertex u.
func (g *Graph) OutEdges(u int) []edge {
	return g.outEdges[u]
}

File: typescript/src/graph.ts
/** Directed graph with non-negative edge weights. Vertices are 0..vertexCount-1. */
export class Graph {
  private _outEdges: [number, number][][] = [];
  private edgeCount = 0;

  constructor(vertexCount: number) {
    this._outEdges = Array.from({ length: vertexCount }, () => []);
  }

  vertexCount(): number {
    return this._outEdges.length;
  }

  getEdgeCount(): number {
    return this.edgeCount;
  }

  addEdge(from: number, to: number, weight: number): void {
    if (weight < 0) throw new Error('edge weights must be non-negative');
    this._outEdges[from].push([to, weight]);
    this.edgeCount++;
  }

  outEdges(u: number): [number, number][] {
    return this._outEdges[u];
  }
}
